shared/components/speaker_layer.cpp

```diff
@@ -0,0 +1,261 @@
+}
+
+float SpeakerLayer::distanceToEdge(float width, float height, float angleDegrees) {
+    // Convert angle from degrees to radians
+    float angleRadians = angleDegrees * M_PI / 180.0f;
+    float a = fabs((height / 2.0f) / cos(angleRadians));
+    float b = fabs((width / 2.0f) / sin(angleRadians));
+    return std::min(a, b);
+}
+
+SpeakerLabelPlacement::SpeakerLabelPlacement() {};
+SpeakerLabelPlacement::~SpeakerLabelPlacement() {};
+
+void SpeakerLabelPlacement::reset() {
+    drawableSpeakers.clear();
+}
+
+void SpeakerLabelPlacement::addSpeaker(const std::string& label, float azimuth) {
+    SpUi spUi{ label, azimuth, azimuth, false };
+    while (spUi.spAz >= 360.0) spUi.spAz -= 360.0;
+    while (spUi.spAz < 0.0) spUi.spAz += 360.0;
+    spUi.labAz = spUi.spAz;
+    drawableSpeakers.push_back(spUi);
+}
+
+void SpeakerLabelPlacement::drawSpeakers(Graphics& g, SpeakerLayer* layer) {
+    sortSpeakers();
+    int lastIndex = static_cast<int>(drawableSpeakers.size()) - 1;
+    // Process speaker positions. Ignore 0 deg and 180 - we always want to leave those as they are.
+    /// Check from 0 to 180 for spacing
+    for (int i = 0; i <= lastIndex; ++i) {
+        const float thisAz = drawableSpeakers[i].spAz;
+        if (thisAz == 0.f) continue;
+        if (thisAz >= 180.f) break;
+        processSpeaker(i);
+    }
+    /// Check from 360 to 180 for spacing
+    for (int i = lastIndex; i >= 0; --i) {
+        const float thisAz = drawableSpeakers[i].spAz;
+        if (thisAz == 0.f) continue;
+        if (thisAz <= 180.f) break;
+        processSpeaker(i);
+    }
+    // Draw them
+    for (auto const& spUi : drawableSpeakers) {
+        layer->drawSpeaker(g, spUi.spAz, spUi.labAz, spUi.label, spUi.inner);
+    }
+}
+
+void SpeakerLabelPlacement::sortSpeakers()
+{
+    std::sort(drawableSpeakers.begin(), drawableSpeakers.end(),
+        [](const SpUi& a, const SpUi& b) {
+            return a.spAz < b.spAz;
+        }
+    );
+}
+
+void SpeakerLabelPlacement::processSpeaker(int spkIndex)
+{
+    if (tooCloseToPrev(spkIndex)) {
+        drawableSpeakers[spkIndex].inner = true;
+        drawableSpeakers[spkIndex].labAz += bigAzNudge;
+        getPrevSpk(spkIndex)->labAz -= littleAzNudge;
+    }
+
+    if (tooCloseToNext(spkIndex)) {
+        drawableSpeakers[spkIndex].inner = true;
+        drawableSpeakers[spkIndex].labAz -= bigAzNudge;
+        getNextSpk(spkIndex)->labAz += littleAzNudge;
+    }
+}
+
+bool SpeakerLabelPlacement::tooCloseToNext(int spkIndex)
+{
+    auto spk = getNextSpk(spkIndex);
+    return tooClose(&drawableSpeakers[spkIndex], spk);
+}
+
+bool SpeakerLabelPlacement::tooCloseToPrev(int spkIndex)
+{
+    auto spk = getPrevSpk(spkIndex);
+    return tooClose(&drawableSpeakers[spkIndex], spk);
+}
+
+bool SpeakerLabelPlacement::tooClose(const SpUi* spkA, const SpUi* spkB)
+{
+    // Check there actually is a neighbour
+    if (!spkA || !spkB) return false;
```

Comment:
I think these should only ever be null if there is only one speaker, so could you move that check into drawSpeakers() and then avoid having to do anything with pointers?
Could change the checks in getNextSpk/getPrevSpk to assertions (or pop one in processSpeaker) to guard against future misuse / make the pre-condition visible?

Reply:
Yeah, would be good to get rid of pointers. I'm reluctant to move the checks too far up the call stack because I think the methods should do something sensible in all cases in case the program flow changes in future (i.e, shouldn't be relying on calling functions to do the checks to make sure the methods will behave as expected).
I'll try figure out a nicer way to do it.